src/adapter/src/util.rs

```diff
@@ -447,31 +447,41 @@ pub(crate) fn viewable_variables<'a>(
         })
 }
 
-/// Verify that the row datums match the expected desc.
-pub fn verify_datum_desc(desc: &RelationDesc, rows: &[Row]) -> Result<(), AdapterError> {
+/// Verify that the rows in [`RowIterator`] match the expected [`RelationDesc`].
+pub fn verify_datum_desc(
+    desc: &RelationDesc,
+    rows: &mut dyn RowIterator,
```

Comment:
Passing this as `mut` so that `peek` can work is unfortunate, since it could potentially mutate current position state incorrectly in the future. Does it make sense to have `offset` be a parameter to a constructor fn that consumes it, then never stores it in the struct?

Reply:
Chatted offline, the [`peek`](https://doc.rust-lang.org/std/iter/struct.Peekable.html#method.peek) method in `std` also requires a `&mut` so we're okay leaving this for now

Reply:
We use a similar pattern in columnation and the likes, but require the iterator to be `Clone` so we can iterate it multiple times. This works because we know that the iterators are cheap to clone, but that might not be the case here. Would it make sense to change the function to take an `Option<&[Datum<'_>]>` instead?